Porting TQL to scalameta 0-1-0-SNAPSHOT

mdemarne opened the pull request:
TQL has been ported to the new version of Scalameta (0-1-0-SNAPSHOT).
Test have been ported to the new version as well but do not pass. Note that it was already the case with Scalameta 0.0.0-M0. A deeper look into TQL is probably required to find the reasons of the fails.

mdemarne committed:
to scalameta 0-1-0-SNAPSHOT

- Changing library dependencies
- Changing nonTriviaFields to fields
- Changing toScalameta to toPTree

Present test do not compile due to quasiquote, not yet implemented for
this version of scalameta.

mdemarne committed:
Correcting examples.

Import of scala.meta._ package must be explicit to use the current
quasiquotes.

mdemarne committed:
$tag -> internaltag, removing Procedures in tests

- Removing procedures from tests (using .def instead)
- Following the renaming of $tag -> internalTag

TODO:
- Change the way trees are converted to scala.meta trees, as ways have been
changed.

mdemarne committed:
Adding proper call to toMTree converter

- Adding proper call to to MTree conveter.
- Correcting macro call to generate traverser ($tag do not exist
anymore).

mdemarne committed:
BF in TraverserHelpers, $tag -> internalTag

Bug Fix. Now everything compile with current scala.meta version.

mdemarne committed:
Forcing to version 0.1.1

Changing version for consistency with scala.meta changes.

mdemarne committed:
Moving TQL to follow same versionning as scalameta

Better to use the same versionning for consistency. Obey will be changed soon
as well, once debugged.

begeric commented:
Hi,
Thanks for porting TQL to the new version of scala.meta.
Considering the tests:
- FusionBenchmarks and OptimizeBenchmarks don't pass because the the file they do the benchmarks on isn't on the repo (I can't put it). Instead use GenSeqLike.scala (or any other file of size ~ 500 LOC or bigger).
- The other tests fail because there seems to be an error in the Traverser (probably in the macro :/). Moreover, the tests in GetValsInMethodsSuite will probably not pass anyway because 'main', 'test' and 'test2' are procedures and not functions. Note that I don't know how procedure are handled in scala.meta right now, but this is something to consider.

Thanks
Eric

mdemarne commented:
Hi Eric,
Thanks for the comments.
- Procedure were temporary removed from Scala.Meta. I assumed that they were translated to Def with a return type of Unit, but I might be wrong. Eugene, can you confirm or infirm?
- Regarding the traverser, I will take a closer look ASAP and see what I can do with the Macro. It will be a good review for me anyway.

The goal here is to include the source into scala.meta, so we don't have to maintain TQL separately.

mdemarne committed:
Changing benchmarks to use GenSeqLike.scala

As benchmarks file cannot be put into git for various reasons. The
line has been commented with a TODO for future usage.

mdemarne commented:
Small information after a few checks:
The `override def traverse(tree: scala.meta.Tree)` method in https://github.com/begeric/TQL-scalameta/blob/master/tqlscalameta/src/main/scala/tools/Traverser.scala#L52 is never called by the ObeyRuleSuite.scala, nor in some cases of GetValsInMethodsSuite.scala. 

That means the problem might come from somewhere else, isn't it?

begeric commented:
That's just what I call the "regular traverser". The one you use to manipulate AST's like with scala.reflect.
TQL makes no use of such a traverser. You should rather look into:
https://github.com/begeric/TQL-scalameta/blob/master/tqlscalameta/src/main/scala/meta/tql/ScalaMetaTraverser.scala

I am not sure that ObeyRuleSuite or GetValsInMethodsSuite make use of a "regular traverser".
Nevertheless, if there is a problem in the traverser generated for TQL there is problably also a problem in the traverser generated for the "regular traverser" (https://github.com/begeric/TQL-scalameta/blob/master/tqlscalametamacros/src/main/scala/tools/traverserBuilder.scala).

xeno-by commented:
1) `getValsInMethods` and friends are failing most probably because the structure of the val/var trees has been changed since 0.0.0-M0: `Defn.Val(_, (b: Term.Name):: Nil,_, _)` has become `Defn.Val(_, Pat.Var.Term(b: Term.Name):: Nil,_, _)`.

2) `listToSetBool` is probably a regression in scalahost. I'll look into it right now.

3) The rest of the errors that I get look like `Could not run test meta.tql.OptimizeBenchmarks: java.lang.ClassCastException: sun.management.GarbageCollectorImpl cannot be cast to javax.management.NotificationEmitter`. Looks like I need to configure something?

xeno-by commented:
2) Yes, indeed that's a change in scalahost. The tree now looks like `Term.Apply(Term.Select(Term.Apply(Term.Name("List")[18], List(Lit.Int(1), Lit.Int(2), Lit.Int(3))), Term.Name("toSet")[19]), List(Lit.Unit(())))`.

xeno-by commented:
2 is a bug, but I don't have time to fix it at the moment. So, I suggest we just change `toSet()` to `toSet(())` and be done with it.

begeric commented:
3) The rest of the errors that I get look like Could not run test meta.tql.OptimizeBenchmarks: java.lang.ClassCastException: sun.management.GarbageCollectorImpl cannot be cast to javax.management.NotificationEmitter. Looks like I need to configure something?

@xeno-by I have no idea what this is, maybe something changed in scalameter. @mdemarne do you have the same problem?
I'll take a look this WE.

mdemarne committed:
Modifying test cases following https://github.com/begeric/TQL-scalameta/pull/2.

- toSet() to toSet(())
- Modifying val traversers to use Pat.

Does not entirely fix the test problems. It seems that nothing is traversed.

mdemarne commented:
1), 2) Modified, but it does not seems to solve the test problems. Perhaps a bug in the traverser generation within the Macro, as Eric suggested. Nothing is ever matched it seems. I will keep looking.
3) No, I do not have this problem - no idea where it might come from... :-/

Thank you both!

begeric commented:
@xeno-by 3) looks like a problem with scalameter and java 1.6. Please see [here](https://github.com/scalameter/scalameter/issues/3).

xeno-by commented:
Allright, I will now be including mdemarne/TQL-scalameta into scalameta/scalameta. Thanks for your help guys!

xeno-by commented:
https://github.com/scalameta/scalameta/pull/113

mdemarne commented:
Following the merging, TQL is now in scalameta. This repository can be discontinued. A new pull request solving TQL traversal problem, as mentioned above, has been added at https://github.com/scalameta/scalameta/pull/117/files. @begeric I would really appreciate if you could just take a quick look at it to say if it's OK, as so far you were the sole author of this code. Thanks.

xeno-by commented:
I think that this repo could stay because of two reasons: 1) development history, referenced by scalameta, 2) not all of the code from here was ported, and some of that stuff might be additionally ported in the future.

mdemarne commented:
Ok, true of course. My bad. I am going to add the changes I did in scalameta to TQL-scalameta in order to be able to close this pull request with the bug fix. Thanks!

begeric commented:
LGTM

mdemarne committed:
Adding case in Builder for Option[Seq[T]].

The traverser prevented any traversing of Templates. This was due to the way
stats are setup into templates as Option[Seq[Stat]], which was not handled
by the previous version of Obey. This is now corrected and the Obey test case
passes as usual.
